fix: allows for schedulings to be defined for operator jobs

shawkins opened the pull request:
closes: #42057

It's probably worth discussing, but we actually have several options here:

- this pr - keep the current behavior that defaults the job scheduling to that of the server, and allow that to be overriden by specific scheduling fields. 

- An alternative to putting the import scheduling on the import cr is to instead put that on the keycloak cr - it's highly likely that users would want the same scheduling for all imports.

- Add a declarative field on or beside the scheduling spec to state whether the scheduling should be inherited - it would default to true as to not be a breaking change, but we could later default it false.

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

shawkins committed:
fix: allows for schedulings to be defined for operator jobs

closes: #42057

Signed-off-by: Steve Hawkins <shawkins@redhat.com>

ryanemerson commented:
I like the separation of scheduling definitions on the two CRs, I think it is the most flexible solution, so +1 to the current approach to the specs.

I am not convinced it makes sense for the import and upgrade jobs to default to the Keycloak scheduling rules though.

Currently the default `TopologySpreadConstraints` prefer that pods are scheduled on distinct zones and nodes. Imagine a K8s cluster with multiple zones, but a single Keycloak instance. I don't think it makes sense that the upgrade/import jobs be provisioned in a separate zone as this will cause additional latency and cost.

I can think of two solutions to prevent ^:

1. Don't define scheduling defaults for these jobs. Simple, but jobs may still be scheduled across zones.

2. Define default scheduling for import and upgrade jobs that prefers co-location via an affinity rule.

I think 2. is the more robust solution. 

Is there any scenario where wouldn't prefer the jobs to be co-located?

shawkins commented:
> Can we add this to notable changes and document that by default import/upgrade jobs will be co-located unless explicitly configured otherwise?

@vmuzikar we didn't explicitly document the that Jobs inherit scheduling from the Keycloak CR, so do you see this as a notable change - or a breaking change that would necessitate a flag in a minor release?

vmuzikar reviewed (dismissed):
@shawkins Thanks, LGTM.

vmuzikar commented:
> but a single Keycloak instance

This would not seem production ready – a single KC pod. :) I doubt that would be too common.

vmuzikar commented:
> Is there any scenario where wouldn't prefer the jobs to be co-located?

In strictly resourced envs where the user doesn't expect more than 1 KC pod on a node.

ryanemerson commented:
> Now, won't it be confusing to the user that suddenly a new "KC pod" appeared in the same AZ when some other AZ has 0 pods?

Presumably the pods have a different naming convention as they're not part of the StatefulSet, so I think that would make the difference clear.

> Additionally, what would be the disadvantage of scheduling the Jobs in a different AZ? Neither the Import Job, nor the Update Job need to communicate with the other KC Pods, no?

That's a good point about the pod communication. When I wrote https://github.com/keycloak/keycloak/pull/42310#issuecomment-3252831668, I had made some incorrect assumptions about how the import job worked. So you're right, network cost for cross-zone traffic is not a concern here.

It's minor, but one disadvantage I can think of is resource allocation. Inheriting the default policies means that more weight will be attached to scheduling the job pod to a distinct AZ/node then a node that has more available resources.

Very much an edge-case, but job execution and simultaneous scaling of Keycloak `spec.instances` could also lead to a scenario where a keycloak server pod is not provisioned on a distinct az. Consider the scenario described in https://github.com/keycloak/keycloak/pull/42310#issuecomment-3265148434 again, if the job is on AZ3, it could be that `keycloak-2` then gets provisioned on AZ1 or AZ2.

> One reason might be the DB latency but that shouldn't be an issue if the scheduling actually allows the KC Pods to be scheduled in various AZs.

+1. If the user wants to co-locate an import job with the DB writer instance, they will now have the flexibility in `keycloakrealmimports` to do that. That will be deployment dependent though, so I don't think we can further improve this in a generic way.

shawkins commented:
Talked about this in our meeting. It seems like the most straight-forward UX is to make our best effort to make things work as much as expected by default and be done in only a single place, so we should not inherit the default scheduling and we'll move the scheduling of import jobs to the keycloak cr.

@vmuzikar @ryanemerson @pruivo we also wanted to confirm there's no chance the switch from non-stretch to stretch clusters could be a breaking change for any existing users?

ryanemerson commented:
> @vmuzikar @ryanemerson @pruivo we also wanted to confirm there's no chance the switch from non-stretch to stretch clusters could be a breaking change for any existing users?

You mean the new default `podTopologySpreadConstraints` vs the old anti-affinity rules? It shouldn't impact functionality in the sense that everything will work as before, but if the user is running a multi-az k8s cluster then pods will now be scheduled on distinct zones opposed to being co-located within the same az.

The changes doc is outdated, as it still talks about anti-affinity rules. If you're happy with above change in behaviour, then I will create a PR to fix it.

shawkins commented:
Updated the pr based upon my last comment.

> You mean the new default podTopologySpreadConstraints vs the old anti-affinity rules? It shouldn't impact functionality in the sense that everything will work as before, but if the user is running a multi-az k8s cluster then pods will now be scheduled on distinct zones opposed to being co-located within the same az.

Just wanted to make sure there wasn't a case to be made for this being a breaking change - this could even be if someone experiences unexpected cross-az data shipment charges that weren't happening before.

> The changes doc is outdated, as it still talks about anti-affinity rules. If you're happy with above change in behaviour, then I will create a PR to fix it.

I had fixed one occurrance of that in this pr, went ahead and addressed the other as well. If need be it can be separated to a different issue / pr.

ryanemerson commented on docs/guides/operator/advanced-configuration.adoc:
```diff
@@ -294,8 +294,8 @@ If you do not specify a custom topology spread constraint, your Pods will have a
 
-Realm import and update Jobs inherit the scheduling of the server Pods. If your server Pod scheduling is highly restrictive this may result in the 
-Job Pods being unschedulable. In this situation you may override the Job scheduling using the KeycloakRealmImport CR `spec.scheduling` and Keycloak CR `spec.update.scheduling` fields.
+Realm import and update Jobs inherit the scheduling of the server Pods if it is specified the Keycloak CR `spec.shceduling` field. If your server Pod scheduling is highly restrictive this may result in the 
```
`s/shceduling/scheduling`

ryanemerson commented on operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java:
```diff
@@ -360,4 +365,12 @@ public void setReadinessProbeSpec(ProbeSpec readinessProbeSpec) {
         this.startupProbeSpec = startupProbeSpec;
     }
+
+    public SchedulingSpec getImportSchedulingSpec() {
```
Does it make sense for us to take this opportunity to nest the import scheduling so we have `spec.import.scheduling`?

The benefits would be:

1. Consistent with `spec.update.scheduling`
2. Allows for future global import configuration that maybe required e.g. `spec.update.<tbdElement>`. I don't have a concrete example, but you never know what the future might bring.

shawkins replied:
> Does it make sense for us to take this opportunity to nest the import scheduling so we have spec.import.scheduling?

It does, we talked about this a little in our team meeting, but I didn't reflect it in this pr. It's possible that users will eventually want to tweak the resource allocation for the update and import jobs as well, so having a separate stanza is good.

ryanemerson reviewed (commented):
> I had fixed one occurrance of that in this pr, went ahead and addressed the other as well. If need be it can be separated to a different issue / pr.

Thanks! It makes sense for it to be addressed here :+1:

shawkins commented:
@ryanemerson should be good now.

ryanemerson reviewed (commented):
Thanks @shawkins, other than a minor docs fix LGTM assuming CI is :green_circle:

shawkins commented on docs/guides/operator/advanced-configuration.adoc:
```diff
@@ -290,7 +290,26 @@ spec:
+
+==== Job Scheduling
+
+Realm import and update Jobs inherit the scheduling of the server Pods if it is specified the Keycloak CR `spec.scheduling` field. If your server Pod scheduling is highly restrictive this may result in the 
```
Corrected, thanks @ryanemerson

shawkins committed:
Update docs/guides/operator/advanced-configuration.adoc

Co-authored-by: Václav Muzikář <vaclav@muzikari.cz>
Signed-off-by: Steven Hawkins <shawkins@redhat.com>

vmuzikar reviewed (approved):
LGTM, thank you @shawkins!
